envoy: Use github.com/cilium/proxy

jrajahalme opened the pull request:
Add a new variable `envoy.RequiredEnvoyVersionSHA` and check that the
running Envoy version matches that instead of the current Cilium SHA.

This requires that Envoy version dependency is manually updated, but
allows Cilium SHA to advance without requiring to rebuild Envoy.

Pull cilium-envoy from docker image `cilium/cilium-envoy:<CILIUM_ENVOY_SHA>`, where `<CILIUM_ENVOY_SHA>` is the git SHA of the dependency to the new github.com/cilium/proxy repo.

Use a new (experimental) builder image from which the Envoy build dependencies have been removed.

Pull in the generated protobuf API files via a vendor dependency to github.com/cilium/proxy. This is also referred to via a SHA, which may be different from the CILIUM_ENVOY_SHA due to the protobufs maybe changing less often.

Proxy dependency to the go build generated `libcilium.h` that captures the exported function prototypes from `proxylib/proxylib.go` is currently handled manually without any checks. This means that if `proxylib/proxylib.go` exported function prototypes are changed Envoy will fail when calling those functions.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6123)
<!-- Reviewable:end -->

jrajahalme committed:
test: Remove Envoy unit tests

Envoy tests are no longer in this repo, but have been moved to
github.com/cilium/proxy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
.gitignore: Remove envoy ignores that no longer apply.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Import github.com/cilium/proxy as a vendor dependency.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
build: Check that Envoy version SHAs match

Add a step that checks that Envoy SHAs in CILIUM_ENVOY_SHA and
pkg/envoy/envoy.go match. By default cilium fails to start if the
versions do not match, try to prevent that from happening by checking
at the build time.

CILIUM_ENVOY_SHA is used as the tag for the cilium/cilium-envoy image,
which should have the same SHA within the Envoy version inormation
cilium is checking on startup.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

tgraf commented on Makefile:
```diff
@@ -124,23 +121,20 @@ install:
 
 docker-image-builder:
-	cd envoy && docker build -t "quay.io/cilium/cilium-builder:$(UTC_DATE)" .
+	docker build -t "cilium/cilium-builder:jarno-test" -f Dockerfile.builder .
```
I assume the `jarno-test` is temporary and will be removed in the final submission?

jrajahalme replied:
Right!

tgraf commented on pkg/envoy/envoy.go:
```diff
@@ -39,6 +39,7 @@ import (
 var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "envoy-manager")
 
 var (
+	RequiredEnvoyVersionSHA = "fd3866831d8f85e27120f34d98133343fc9c4e0e"
```
Please add a comment that describes when to update this.

jrajahalme replied:
Done.

tgraf commented on test/docker-compose.yml:
```diff
@@ -12,7 +12,7 @@ services:
     privileged: true
   base_image:
-    image: "quay.io/cilium/cilium-builder:2018-10-29"
+    image: "cilium/cilium-builder:jarno-test"
```
Temporary?

jrajahalme replied:
Yes, was testing with a stripped down builder. Can either produce official builder with the new Dockerfile, or use the old ones for now.

tgraf commented on Gopkg.toml:
```diff
@@ -298,3 +298,7 @@ required = [
   version = "1.0.0"
+
+[[constraint]]
+  name = "github.com/cilium/proxy"
```
Nice!

tgraf reviewed (changes requested):
Awesome!

aanm commented on Dockerfile:
```diff
@@ -8,7 +17,7 @@
 # that are not backwards compatible.
 #
-FROM quay.io/cilium/cilium-builder:2018-10-29 as builder
+FROM cilium/cilium-builder:jarno-test as builder
```
Change image tag

jrajahalme replied:
Built the new builder image in quay.io, with tag 2018-11-01. It is now only 235MB ;-)

aanm commented on Dockerfile.builder:
```diff
@@ -0,0 +1,52 @@
+#
```
Can you leave a description where is this image used?

jrajahalme replied:
Added comment:
```
#
# Cilium build-time dependencies.
# Image created from this file is used to build Cilium.
#
```

Note that while I removed the most obvious Envoy dependencies from here, I'm not sure if all that remains is really needed to build Cilium. Maybe you can have look?

jrajahalme replied:
@aanm ^^^^

aanm replied:
I know it's used to build Cilium :-) I'm just not sure if this one is the cilium/builder or the cilium/runtime

aanm commented on Documentation/contributing.rst:
```diff
@@ -41,22 +41,10 @@ contribute to Cilium:
 +----------------------------------------------------------------------------------+--------------------------+-------------------------------------------------------------------------------+
 + `Docker-Compose <https://docs.docker.com/compose/install/>`_                     | OS-Dependent             | N/A (OS-specific)                                                             |
 +----------------------------------------------------------------------------------+--------------------------+-------------------------------------------------------------------------------+
-+ `Cmake  <https://cmake.org/download/>`_                                          | OS-Dependent             | N/A (OS-specific)                                                             |
```
:tada:

aanm commented on test/run-envoy-unit-tests.sh:
```diff
@@ -1,5 +0,0 @@
-
-set -e
-
-vagrant ssh runtime -c "cd /home/vagrant/go/src/github.com/cilium/cilium ;make build; make tests-envoy"
```
How will we test envoy from now on? Is there a plan to create a jenkins job for the new repo? cc @eloycoto

jrajahalme replied:
No concrete plan yet, but obviously needs to be done.

jrajahalme replied:
So far there are no code changes in the cilium/proxy repo, so we are good.

aanm commented on Makefile:
```diff
@@ -121,7 +121,9 @@ install:
 	echo "$(GIT_VERSION)" >GIT_VERSION
 
-docker-image: clean GIT_VERSION
+docker-image: clean docker-image-no-clean
```
This can be made in a different PR. I think @nebril wants this a lot.

jrajahalme replied:
Done & merged (#6132).

aanm reviewed (changes requested):
:tada: :tada: :tada:

jrajahalme committed:
envoy: Make Envoy dependency version explicit.

Add a new variable `envoy.RequiredEnvoyVersionSHA` and check that the
running Envoy version matches that instead of the current Cilium SHA.

This requires that Envoy version dependency is manually updated, but
allows Cilium SHA to advance without requiring to rebuild Envoy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
